enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+    - required: ["grpcServiceSource"]
+    - required: ["configMapSource"]
+                  
+``` 
+
+This option introduces breaking changes. This means that while we are on the depracation path, the users will be able to configure their `CatalogSource`s in the current way, _OR_, use one of `grpcPodSource`, `grpcServiceSource`, or `configMapSource` (with precendence given to the new configuration). 
+
+**Pros**
+  1. Each CatalogSource type has its own configuration structure
+  2. We don't polute the top level with endless optional attributes that behave differently depending on the values of other attributes
+  3. Better UX
+  4. New catalog source types can be added easily
+
+
+**Cons**
+  1. Introduces breaking changes
+  2. There is not first-class support for union type validation in kubebuilder/operator-sdk. We'll need another mechanism to add and test these validations (e.g. patch that gets executed after `make controller-gen` in the `api` repo)
+
+#### CatalogSourceClass
+
+In this option, we transion the `CatalogSource` to a more plugable architecture by introducing a `catalogSourceClass` and a schemaless `parameters` block where arbitrary parameters can be passed in. The controller would then be in charge of reconciling catalog sources of classes it recognizes.
+
+In the future, we'd have something like this:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha2
+kind: CatalogSource
+metadata:
+  name: operatorhubio-catalog
+  namespace: olm
+spec:
+  catalogSourceClass: grpc-pod
+  displayName: Community Operators
+  publisher: OperatorHub.io
+  parameters:
+    image: quay.io/operatorhubio/catalog:latest
+    # additional parameters are options and override the underlying podspec
+```
+
+**Pros**
+  1. Extensible architecture: users can provide their own `CatalogSourceClass` types and controllers
+  2. Better UX than the status quo: top level attributes are shared and `parameters` should only contain information relevant for that class of catalog source
+
+**Cons**
+  1. `parameters` is schemaless. So, we'd also need an adminission webhook
+  2. Requires deprecation
+
+#### ImageConfig
+
+This option is the least invasive of the three. Here we add an additional optional attribute `ImageConfig` that contains the parts of the pod spec we'd like to override. For example:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: operatorhubio-catalog
+  namespace: olm
+spec:
+  sourceType: grpc
+  image: quay.io/operatorhubio/catalog:latest
+  imageConfig:
+    nodeSelector:
+        node-role.kubernetes.io/master: ""
+    tolerations:
+    - key: "node-role.kubernetes.io/master"
+      operator: Exists
+      effect: "NoSchedule"
+  displayName: Community Operators
+  publisher: OperatorHub.io
+```
+
+We could define `imageConfig` to be a `corev1.PodSpec` and override any and all attributes specified. We could also keep the definition minimal and slowly add additional fields as they are needed.
```

Comment:
In the `corev1.PodSpec` case, we'd run into the problem that the pod spec _also_ contains an image field, so it would be an edge case for the controller to handle and a potential point of confusion for users.

Another consideration of `corev1.PodSpec` is that its a potential privilege escalation vulnerability. It would allow anyone with `CatalogSource` write access to create arbitrary pods on the cluster with the privileges of the catalog operator.

Reply:
I think the question of what to expose is super relevant. Once we sort out which path to take, this will be the next logical questions. IMO, we should keep it as small as possible. Could be just image, tolerations and node-selector, but we could also see if there are other aspects that could make sense too. At the end of the day, most of the control should be on the controller to say what the pod executes, etc. Maybe scheduling is the only concern the admin might want to override? Maybe additional labels and annotations as well? But that should be about it. At least to start with.

Reply:
Great call out with the security stuff tho. That's absolutely right!

Reply:
Yeah, this is just `podPresets` all over again, which we've already debated a number of times and already died in Kubernetes. https://v1-18.docs.kubernetes.io/docs/concepts/workloads/pods/podpreset/

+1 to Joe, let's just leave this as a static set of fields that can be set. I don't see a reason why we would want to expand this proposal to do more than the subscription config can already do, unless there's a specific need for something else.